### DISCUSSION TITLE
Is there an OSD (special) or an idea to create one (or one of the NWG tools has the ability to act as one)?

### DISCUSSION THREAD

**Discussion by msmafra** [nwg-piotr/nwg-shell#450] [Ideas]
Hi!
This is probably just me trying to almost a DE from Hyprland.
It came to mind when testing nwg-displays, my actual idea is not a feature focus OSD like they normally are, it would be more like a notification accepting any text input from a script too.
For example I could monitor Hyprlanc IPC socket with a script and when I change to a workspace it would show its number and or name or even add the monitor on the screen; or I just enabled a submap and while I'm on that "mode" it would show the submap name on the screen. I do similar but not on any place of the screen with Waybar where I make it blink too.

...

I tried some weeks back with the Hyprland notification system, but it doesn't allow position and formatting beyond some colors and font related things.

**Comment by nwg-piotr** [comment]
Maybe you could try with swaync?

Actually it would be quite easy to code something like this. I only need a possible application within nwg-shell.

**Reply by msmafra** [reply]
I barely finished configuring my swaync. I don't know its capabilities yet.

**Comment by nwg-piotr** [comment]
Is it what you want?

https://youtu.be/Imro6l0xqo4

**Reply by msmafra** [reply]
> Is it what you want?
> 
> https://youtu.be/Imro6l0xqo4

Yes, that seems to be it. I'm confused, you made one that quick?!

**Reply by nwg-piotr** [reply]
I said it's easy. ;) Some features (window positioning, optional config file) are still missing, but this shouldn't take much time.

**Comment by nwg-piotr** [comment]
OK, it took [less than 300 lines of code](https://github.com/nwg-piotr/nwg-shell-config/blob/hud/nwg_shell_config/hud.py). I need to test it better tomorrow. The `nwg-hud` script will be a part of the nwg-shell-config package.

**Reply by msmafra** [reply]
> OK, it took [less than 300 lines of code](https://github.com/nwg-piotr/nwg-shell-config/blob/hud/nwg_shell_config/hud.py). I need to test it better tomorrow. The `nwg-hud` script will be a part of the nwg-shell-config package.

I'm testing it, probably shouldn't yet and probably in the wrong way. But I noticed to things, testing via my bash script, it has a delay and another thing it interferes with focus. Only really noticed when I set it to 3000+ timeout. I takes focus from the window while is on the screen.

```
#!/usr/bin/env bash
workspace_hud() {

	local SCRIPT_NAME
	local SCRIPT_FOLDER
	local -a workspace_types
	local -i TIMEOUT
	local workspace_name
	SCRIPT_NAME="$(basename "${0}")"
	SCRIPT_FOLDER="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
	workspace_types=(
		"workspace>>"
		"activespecial>>"
	)
	TIMEOUT=500

	if [[ ${1:0:11} == "${workspace_types[0]}" ]]; then

		# python3 "${SCRIPT_FOLDER}/hud.py" -t "${TIMEOUT}" -m " Switched to ${1//[^[:digit:]]/}" -i ~/.config/hypr/wlogout/icons/hibernate-new.png
		python3 "${SCRIPT_FOLDER}/hud.py" -t "${TIMEOUT}" -m "  Workspace ${1//[^[:digit:]]/}"

	elif [[ ${1:0:15} == "${workspace_types[1]}" ]]; then

		workspace_name="$(printf "%s" "${1}" | xargs -0 | grep -Po "(?<=:)(.*?)(?=,)")"
		python3 "${SCRIPT_FOLDER}/hud.py" -t "${TIMEOUT}" -m "  Workspace ${workspace_name}"

	else

		echo "nothing to do"

	fi

}

socat - "UNIX-CONNECT:$XDG_RUNTIME_DIR/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock" | while read -r line; do
	workspace_hud "${line}"
done
```

**Comment by nwg-piotr** [comment]
I'm useless at bash scripts. For a sample usage see https://github.com/nwg-piotr/nwg-shell-config#nwg-hud.

**Reply by msmafra** [reply]
> I'm useless at bash scripts. For a sample usage see https://github.com/nwg-piotr/nwg-shell-config#nwg-hud.

That script is based on the one to expand Hyprland's functionality via sockets/IPC.
The COPR for Fedora is not updated yet, so I copied the .py and other files to try to use/test it.

And thank you kindly for that. :)

**Reply by nwg-piotr** [reply]
Hope it's going to work well. Honesty, I could've copied some imported functions into the script, in order not to make you use external files. Didn't know you were on Fedora.

**Comment by msmafra** [comment]
It's already working very well. Thank, thank you kindly for that. 
I use the script because setting it to run in the background, allows for it to show in any method of switching to workspace, being it key binds, window rules, workspaces rules or by click on the workspace button. I'll experiment with it for more things that, at least for me, notifications would be weird like when in Submap mode.

But,

It just has a delay that I thought could be from the usage of my script but using the method you documented it also has a delay before showing. It waits for the workspace be fully on the screen, so it shows, I'm not sure if my animations, if a Hyprland thing interfering with it.
Using it from inside my hypr/scripts folder, because I copied config.json and style.css to a hud subfolder, it works directly, but  if I copy it to, for example, ~/.local/bin it looks for a style file inside go's folder (??).
```
nwg-hud -m "  WS 1" -t 5000
gtk-css-provider-error-quark: <broken file>:1:0Failed to import: Erro ao abrir arquivo /home/coop/go/bin/hud/style.css: No such file or directory (2)
```

**Reply by nwg-piotr** [reply]
Well, the path depends on what the `get_config_home()` function returns. And the function is being imported from tools.py. Have you copied it into the script?

**Reply by msmafra** [reply]
No, I just copied hud.py, style.css and config.json.

**Reply by nwg-piotr** [reply]
Actually I have no idea where does it take the `/home/coop/go/bin/hud/style.css` path from. Hopefully Fedora updates the nwg-shell-config package soon, then you'll have the `nwg-hud` command built in.

**Reply by msmafra** [reply]
It's up to date now, today it updated, but unfortunately it only works if I install the nwg-shell package with all the dependencies, installing only nwg-shell-config doesn't work, but.
I keep using the way I've been using it and let it be. Thanks!
Sorry to keep bothering you with this.

```
❯ nwg-hud -m "  WS 1" -t 5000
Traceback (most recent call last):
  File "/usr/bin/nwg-hud", line 33, in <module>
    sys.exit(load_entry_point('nwg-shell-config==0.5.54', 'gui_scripts', 'nwg-hud')())
             ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/bin/nwg-hud", line 25, in importlib_load_entry_point
    return next(matches).load()
           ~~~~~~~~~~~~~~~~~~^^
  File "/usr/lib64/python3.13/importlib/metadata/__init__.py", line 179, in load
    module = import_module(match.group('module'))
  File "/usr/lib64/python3.13/importlib/__init__.py", line 88, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 1022, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/usr/lib/python3.13/site-packages/nwg_shell_config/hud.py", line 33, in <module>
    from psutil import process_iter
ModuleNotFoundError: No module named 'psutil'
```

**Reply by nwg-piotr** [reply]
Just install `python-psutil`. :D
